AX: Fix safer-cpp violations for WebAccessibilityObjectWrapperMac

twilco opened the pull request:
#### 8e1385e99bc38eef78ab055b746d9afccce1e78b
<pre>
AX: Fix safer-cpp violations for WebAccessibilityObjectWrapperMac
<a href="https://bugs.webkit.org/show_bug.cgi?id=297606">https://bugs.webkit.org/show_bug.cgi?id=297606</a>
<a href="https://rdar.apple.com/158699090">rdar://158699090</a>

Reviewed by Joshua Hoffman.

Apply the safer CPP rules to this file.

<a href="https://github.com/WebKit/WebKit/wiki/Safer-CPP-Guidelines">https://github.com/WebKit/WebKit/wiki/Safer-CPP-Guidelines</a>

* Source/WebCore/SaferCPPExpectations/RetainPtrCtorAdoptCheckerExpectations:
* Source/WebCore/SaferCPPExpectations/UncheckedCallArgsCheckerExpectations:
* Source/WebCore/SaferCPPExpectations/UncheckedLocalVarsCheckerExpectations:
* Source/WebCore/SaferCPPExpectations/UncountedCallArgsCheckerExpectations:
* Source/WebCore/SaferCPPExpectations/UncountedLambdaCapturesCheckerExpectations:
* Source/WebCore/SaferCPPExpectations/UncountedLocalVarsCheckerExpectations:
* Source/WebCore/SaferCPPExpectations/UnretainedCallArgsCheckerExpectations:
* Source/WebCore/SaferCPPExpectations/UnretainedLambdaCapturesCheckerExpectations:
* Source/WebCore/SaferCPPExpectations/UnretainedLocalVarsCheckerExpectations:

Canonical link: <a href="https://commits.webkit.org/299522@main">https://commits.webkit.org/299522@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/1a1a334a3af6e4ed6a57918a2f6e8d0d685802eb

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/119244 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/38927 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/29581 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/125475 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/71309 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/c8cce401-69cc-44a2-ac84-0681b83a8dca) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/121122 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/39623 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/47508 "Built successfully") | [❌ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/90595 "Found 1 new test failure: imported/w3c/web-platform-tests/css/css-multicol/crashtests/inline-float-parallel-flow.html (failure)") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/59954 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/f2c8b9ea-17f2-47d6-8a24-abc351e6b252) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/122197 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/31600 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/106900 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/71009 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/ebd2911a-9f9c-4164-8d1e-d83fef6f566c) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/30643 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/25013 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/69125 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/101047 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/25198 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/128488 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/46156 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/34892 "Found 1 new test failure: http/tests/websocket/web-socket-loads-captured-in-per-page-domains.html (failure)") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/99160 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/46521 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/103110 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/98938 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/25156 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/44404 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/22410 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/42729 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/46022 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/51706 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/45488 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/48838 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/47178 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

hoffmanjoshua commented on Source/WebCore/accessibility/mac/WebAccessibilityObjectWrapperMac.mm:
```diff
@@ -2209,55 +2219,49 @@ - (NSArray *)accessibilityParameterizedAttributeNames
-
-    if (textParamAttrs == nil) {
-        auto tempArray = adoptNS([[NSMutableArray alloc] initWithArray:paramAttrs]);
+    static NeverDestroyed<RetainPtr<NSArray>> paramAttrs = @[
```
The old implementation checked for `nil` to prevent creating this array after it has been initialized. Does using `NeverDestroyed` accomplish the same thing?

twilco replied:
Function-local statics are guaranteed to be only initialized once, so we should be good here 🙂

hoffmanjoshua commented on Source/WebCore/accessibility/mac/WebAccessibilityObjectWrapperMac.mm:
```diff
@@ -2900,7 +2909,7 @@ static bool isMatchingPlugin(AXCoreObject& axObject, const AccessibilitySearchCr
 }
 
-static NSRect computeTextBoundsForRange(NSRange range, const AXCoreObject& backingObject)
+- (NSRect)computeTextBoundsForRange:(NSRange)range backingObject:(const AXCoreObject&)backingObject
```
I wonder if we can remove the `backingObject` parameter here, and just get it from `self` as needed (which you do below in `retrieveValueFromMainThread`), not that this is not a static method?

twilco replied:
For performance, it's better to pass the `backingObject` around where we can. This also guarantees that one single accessibility request operates on the same backing object, guarding against the case where a wrapper switches its backing object mid-request, which technically shouldn't happen...but we've had bugs from this in the past.

hoffmanjoshua replied:
Good point, makes sense then!

hoffmanjoshua commented on Source/WebCore/platform/Widget.cpp:
```diff
@@ -358,6 +358,13 @@ FloatRect Widget::convertFromContainingView(const FloatRect& parentRect) const
 // MARK: -
 
+#if PLATFORM(COCOA)
+RetainPtr<PlatformWidget> Widget::protectedPlatformWidget() const
```
Do you think we should inline this?

twilco replied:
This would be a great candidate for that indeed, but unfortunately there is no `WidgetInlines.h` file, so we'd need to create that. Could definitely do this in a future change!

hoffmanjoshua reviewed (approved):
Just a few questions, but looks good to me!

twilco committed:
AX: Fix safer-cpp violations for WebAccessibilityObjectWrapperMac
https://bugs.webkit.org/show_bug.cgi?id=297606
rdar://158699090

Reviewed by Joshua Hoffman.

Apply the safer CPP rules to this file.

https://github.com/WebKit/WebKit/wiki/Safer-CPP-Guidelines

* Source/WebCore/SaferCPPExpectations/RetainPtrCtorAdoptCheckerExpectations:
* Source/WebCore/SaferCPPExpectations/UncheckedCallArgsCheckerExpectations:
* Source/WebCore/SaferCPPExpectations/UncheckedLocalVarsCheckerExpectations:
* Source/WebCore/SaferCPPExpectations/UncountedCallArgsCheckerExpectations:
* Source/WebCore/SaferCPPExpectations/UncountedLambdaCapturesCheckerExpectations:
* Source/WebCore/SaferCPPExpectations/UncountedLocalVarsCheckerExpectations:
* Source/WebCore/SaferCPPExpectations/UnretainedCallArgsCheckerExpectations:
* Source/WebCore/SaferCPPExpectations/UnretainedLambdaCapturesCheckerExpectations:
* Source/WebCore/SaferCPPExpectations/UnretainedLocalVarsCheckerExpectations:

Canonical link: https://commits.webkit.org/299522@main
